src/metatrain/experimental/nanopet/model.py

```diff
@@ -0,0 +1,474 @@
+        for target_name, target_info in dataset_info.targets.items():
+            self._add_output(target_name, target_info)
+
+        self.register_buffer(
+            "species_to_species_index",
+            torch.full(
+                (max(self.atomic_types) + 1,),
+                -1,
+            ),
+        )
+        for i, species in enumerate(self.atomic_types):
+            self.species_to_species_index[species] = i
+
+        # additive models: these are handled by the trainer at training
+        # time, and they are added to the output at evaluation time
+        composition_model = CompositionModel(
+            model_hypers={},
+            dataset_info=dataset_info,
+        )
+        additive_models = [composition_model]
+        if self.hypers["zbl"]:
+            additive_models.append(ZBL(model_hypers, dataset_info))
+        self.additive_models = torch.nn.ModuleList(additive_models)
+
+        # cache keys, components, properties labels
+        self.single_label = Labels.single()
+        self.key_labels = {
+            output_name: copy.deepcopy(dataset_info.targets[output_name].layout.keys)
+            for output_name in self.outputs.keys()
+            if "mtt::aux::" not in output_name
+        }
+        self.component_labels = {
+            output_name: copy.deepcopy(
+                dataset_info.targets[output_name].layout.block().components
+            )
+            for output_name in self.outputs.keys()
+            if "mtt::aux::" not in output_name
+        }
+        self.property_labels = {
+            output_name: copy.deepcopy(
+                dataset_info.targets[output_name].layout.block().properties
+            )
+            for output_name in self.outputs.keys()
+            if "mtt::aux::" not in output_name
+        }
+
+    def restart(self, dataset_info: DatasetInfo) -> "NanoPET":
+        # merge old and new dataset info
+        merged_info = self.dataset_info.union(dataset_info)
+        new_atomic_types = [
+            at for at in merged_info.atomic_types if at not in self.atomic_types
+        ]
+        new_targets = {
+            key: value
+            for key, value in merged_info.targets.items()
+            if key not in self.dataset_info.targets
+        }
+
+        if len(new_atomic_types) > 0:
+            raise ValueError(
+                f"New atomic types found in the dataset: {new_atomic_types}. "
+                "The nanoPET model does not support adding new atomic types."
+            )
+
+        # register new outputs as new last layers
+        for target_name, target in new_targets.items():
+            self._add_output(target_name, target)
+
+        self.dataset_info = merged_info
+        self.atomic_types = sorted(self.atomic_types)
+
+        return self
+
+    def forward(
+        self,
+        systems: List[System],
+        outputs: Dict[str, ModelOutput],
+        selected_atoms: Optional[Labels] = None,
+    ) -> Dict[str, TensorMap]:
```

Comment:
could you maybe add somewhere an architecture explanation + link to the PET paper? This would help a lot to understand the code without already knowing PET

Reply:
The architecture explanation is impossible... I would have to write half a page of comments and still nobody would understand. I will add the link though

Reply:
any kind of high level overview would help here, and make sure others can also help maintain this code as needed, instead of requiring you to do any change.